### PR TITLE
Registration Revision Views

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Purpose

To execute the API contract for schema versioning and write tests to ensure good user and developer experence.

## Changes

- adds top level schema response endpoints (list and detail)
- adds registration relationship endpoint
- adds create and update view level functionality
- adds tests

## QA Notes

Please make verification statements inspired by your code and what your code touches.
- Verify
- Verify

What are the areas of risk?

Any concerns/considerations/questions that development raised?

### CODE REVIEW

File: osf/models/node.py [hunk 1]
```diff
@@ -370,6 +370,19 @@ class AbstractNode(DirtyFieldsMixin, TypedModel, AddonModelMixin, IdentifierMixi
     custom_storage_usage_limit_public = models.DecimalField(decimal_places=9, max_digits=100, null=True, blank=True)
     custom_storage_usage_limit_private = models.DecimalField(decimal_places=9, max_digits=100, null=True, blank=True)
 
+    all_responses = GenericRelation('osf.SchemaResponses')
+
+    @property
+    def schema_responses(self):
```
Comment (jwalz) [inline osf/models/node.py:376]: this is what the `all_responses` GenericRelation does. The fact that that wasn't clear makes me wonder: should we rename that? `all_schema_responses`, maybe? Is confusing because there's two levels of plurality :/

Comment (Johnetordoff) [reply osf/models/node.py:376]: Well maybe just schema responses then? It's going to appear as `schema_reponses.all()` anyway right?

Comment (jwalz) [reply osf/models/node.py:376]: legit, I'll make that change in the other PR.

Any other name changes we want to explore now that we've started writing code around these models?

Comment (Johnetordoff) [reply osf/models/node.py:376]: The whole revision wording, I would call everything schema_responses, but I just followed the API contract. I would change it, but lets check with the FE.

Comment (Johnetordoff) [reply osf/models/node.py:376]: Also `justification` on the model should be `revision_justification`  to match the API contract.

File: api/registrations/views.py [hunk 2]
```diff
@@ -850,3 +853,44 @@ def get_default_queryset(self):
...
+    serializer_class = SchemaResponsesListSerializer
+
+    def get_default_queryset(self):
+        return self.get_node().schema_responses.all()
+
+    def get_queryset(self):
+        return self.get_queryset_from_request()
+
+
+class RegistrationSchemaResponsesDetail(JSONAPIBaseView, generics.RetrieveUpdateDestroyAPIView):
```
Comment (jwalz) [inline api/registrations/views.py:880]: We don't need this one. FE will use the top-level detail view for revision details

File: api/revisions/serializers.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        },
+    )
+
+    registration = RelationshipField(
+        related_view='registrations:registration-detail',
+        related_view_kwargs={'node_id': '<parent._id>'},
+        read_only=True,
+    )
+
+    registration_schema = RelationshipField(
```
Comment (jwalz) [inline api/revisions/serializers.py:37]: We had proposed having this as a FK directly on the SchemaResponses, and I think that's the right call. We'll want access to it for creating/validating ResponseBlocks, so it makes sense for the SchemaResponses to be aware of it instead of inheriting it from the parent. Also, that lets us attach SchemaResponses to entities that may not already have a "schema" field defined.

File: api/revisions/views.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+    serializer_class = SchemaResponsesDetailSerializer
+    view_category = 'schema_responses'
+    view_name = 'schema-responses-detail'
+
+    def get_object(self):
+        return SchemaResponses.objects.get(_id=self.kwargs['revision_id'])
+
+
+class SchemaResponsesActionList(JSONAPIBaseView, ListFilterMixin, generics.ListCreateAPIView):
```
Comment (jwalz) [inline api/revisions/views.py:48]: Let's hold off on the actions stuff for now. We'll implement that once we have the internals of review figured out.

File: osf_tests/factories.py [hunk 5]
```diff
@@ -1115,3 +1115,54 @@ class Meta:
...
+                'title': 'Maybe a section heading',
+                'description': 'Maybe a question label.',
+                'help': 'help text',
+                'required': True
+            }]
+        }]
+    }
+
+    @classmethod
+    def _create(cls, *args, **kwargs):
```
Comment (jwalz) [inline osf_tests/factories.py:1160]: add some logic here to make this a singleton, since it's a canned default?

File: website/templates/project/modal_add_pointer.mako [hunk 6]
```diff
@@ -10,7 +10,7 @@
 
             <div class="modal-body">
 
-                <form role='form'>
+                <form>
```
Comment (jwalz) [inline website/templates/project/modal_add_pointer.mako:13]: looks like this change sneaked in

**Review by jwalz — CHANGES_REQUESTED** [review]

File: api/registrations/views.py [hunk 7]
```diff
@@ -850,3 +856,32 @@ def get_default_queryset(self):
...
+        AdminContributorOrPublic,
+    )
+
+    view_category = 'registrations'
+    view_name = 'schema-responses-list'
+
+    serializer_class = SchemaResponsesListSerializer
+
+    def create(self, request, *args, **kwargs):
+        serializer = self.get_serializer(data=json.loads(self.request._request._body))
```
Comment (Johnetordoff) [inline api/registrations/views.py:877]: I could not find a way to pass this data without overriding this method and shoving the data in there.

File: api/revisions/serializers.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+    def get_is_pending_current_user_approval(self, obj):
+        # TBD
+        return False
+
+
+class SchemaResponsesListSerializer(SchemaResponsesSerializer):
+
+    # overrides Serializer
+    def is_valid(self, clean_html=True, **kwargs):
```
Comment (Johnetordoff) [inline api/revisions/serializers.py:85]: I'm not sure why this is necessary either, but it won't pick up these fields unless I do this.

File: osf/models/schema_responses.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE)
+    parent = GenericForeignKey('content_type', 'object_id')
+
+    schema_responses = models.ManyToManyField(SchemaResponseBlock, related_name='schema_response')
+
+    revision_justification = models.CharField(max_length=2048, null=True)
+    initiator = models.ForeignKey('OSFUser', null=False)
+    submitted_timestamp = NonNaiveDateTimeField(null=True)
+
+    reviews_state = models.CharField(
```
Comment (Johnetordoff) [inline osf/models/schema_responses.py:22]: This is just to practice filtering

File: osf_tests/factories.py (hunk 5 above)
Comment (Johnetordoff) [reply osf_tests/factories.py:1160]: I don't really understand what you're asking.

File: api/registrations/views.py (hunk 7 above)
Comment (jwalz) [reply api/registrations/views.py:877]: We probably don't need this function. I'm okay with saying we should only be POSTing to the top-level endpoint for SchemaResponses.

File: api/revisions/serializers.py [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+
+    def get_revised_responses(self, obj):
+        previous_version = obj.previous_version
+
+        if not previous_version:
+            return []
+        else:
+            return obj.response_blocks.values_list('schema_key', flat=True)
+
+    def update(self, revision, validated_data):
```
Comment (jwalz) [inline api/revisions/serializers.py:122]: Can you leave this unimplemented for now? I have strong enough thoughts that I'd like this and the nuances of the create functionality can be their own PR that we can pair on.

File: api/revisions/serializers.py (hunk 3 above)
Comment (jwalz) [reply api/revisions/serializers.py:48]: Added schema to the SchemaResponses model, so this can just point to schema._id now instead of parent.registered_schema_id

**Review by jwalz — CHANGES_REQUESTED** [review]

File: api/revisions/serializers.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+            )
+        else:
+            raise NotImplementedError()
+
+        return schema_response
+
+
+class SchemaResponsesDetailSerializer(SchemaResponsesSerializer):
+
+    versions = RelationshipField(
```
Comment (jwalz) [inline api/revisions/serializers.py:99]: Not seeing where this is called for. Don't think we need it.

File: api_tests/revisions/views/test_schema_response_list.py [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    def user(self):
+        return AuthUserFactory()
+
+    @pytest.fixture()
+    def payload(self, node):
+        return {
+            'data': {
+                'type': 'schema_responses',
+                'attributes': {
+                    'title': 'new title'
```
Comment (jwalz) [inline api_tests/revisions/views/test_schema_response_list.py:29]: Do schema_responses have a title?

File: api_tests/revisions/views/test_schema_response_list.py [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+                },
+                'relationships': {
+                    'node': {
```
Comment (jwalz) [inline api_tests/revisions/views/test_schema_response_list.py:32]: Contract is expecting "registration" here. Is this just demoing that we can use arbitrary types?

File: api_tests/revisions/views/test_registrations_schema_responses_list.py [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+    @pytest.fixture()
+    def registration(self, schema):
+        return RegistrationFactory()
+
+    @pytest.fixture()
+    def payload(self, registration):
+        return {
+            'data':
+                {
+                    'type': 'registrations',
```
Comment (jwalz) [inline api_tests/revisions/views/test_registrations_schema_responses_list.py:41]: This should be "revisions", the type of the relationship entity should be "registrations"

File: api_tests/revisions/views/test_schema_response_list.py [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+    def node(self, schema):
+        registration = RegistrationFactory()
+        registration.registered_schema.add(schema)
+        registration.save()
+        return registration
+
+    @pytest.fixture()
+    def schema_response(self, user, node, schema):
+        content_type = ContentType.objects.get_for_model(node)
+        return SchemaResponsesFactory(content_type=content_type, object_id=node.id, initiator=node.creator)
```
Comment (jwalz) [inline api_tests/revisions/views/test_schema_response_list.py:56]: can just use `parent=` and Django will figure out the ContentType stuff for you

File: api_tests/revisions/views/test_schema_response_list.py [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+        }
+
+    @pytest.fixture()
+    def schema(self):
+        return RegistrationSchemaFactory()
+
+    @pytest.fixture()
+    def node(self, schema):
+        registration = RegistrationFactory()
+        registration.registered_schema.add(schema)
```
Comment (jwalz) [inline api_tests/revisions/views/test_schema_response_list.py:49]: RegistrationFactory assigns a schema by default. Use `schema=` in the factory call if you want to set this.

File: api_tests/revisions/views/test_schema_response_list.py [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+    def test_schema_response_list(self, app, schema_response, schema_response2, user, url):
+        resp = app.get(url, auth=user.auth)
+        assert resp.status_code == 200
+        data = resp.json['data']
+
+        assert len(data) == 2
+        assert schema_response._id == data[0]['id']
+        assert schema_response2._id == data[1]['id']
+
+    def test_schema_response_create(self, app, node, user, payload, url):
```
Comment (jwalz) [inline api_tests/revisions/views/test_schema_response_list.py:76]: Again, demoing busted permissions

File: api_tests/revisions/views/test_schema_response_detail.py [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+
+    def test_schema_response_detail(self, app, schema_response, user, url):
+        resp = app.get(url, auth=user.auth)
+        assert resp.status_code == 200
+        data = resp.json['data']
+        assert data['id'] == schema_response._id
+        assert data['attributes']['revision_justification'] == schema_response.revision_justification
+        assert data['attributes']['revision_response'] == [{'q1': {'value': 'initial value'}}, {'q2': {'value': 'initial value'}}]
+
+    def test_schema_response_detail_update(self, app, schema_response, payload, user, url):
```
Comment (jwalz) [inline api_tests/revisions/views/test_schema_response_detail.py:105]: moar busted permissions

File: osf/migrations/0236_auto_20210818_1842.py [hunk 19]
```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
```
Comment (jwalz) [inline osf/migrations/0236_auto_20210818_1842.py:1]: please delete this and grab the latest version of the model PR

File: osf/models/schema_responses.py (hunk 9 above)
Comment (jwalz) [reply osf/models/schema_responses.py:21]: Can you omit this and save those tests for later? This is gonna create inconsistency in our migrations

File: osf_tests/factories.py (hunk 5 above)
Comment (jwalz) [reply osf_tests/factories.py:1160]: This is a canned schema. Every call to this factory will produce the same thing, minus the ID. You can safely check to see if a schema already exists with `schema['title']` and just return that schema if it exists instead of creating duplicates.

In general, this feels more like a `DefaultTestSchema` that maybe belongs in.a shared fixture than a `SchemaFactory`

**Review by jwalz — CHANGES_REQUESTED** [review]
Permissions stuff can maybe wait for a follow-on PR if we've got a solution for merging this in the interim. Let's talk about feature branch vs flagging as part of our pairing time?

File: api_tests/revisions/views/test_registrations_schema_responses_list.py (hunk 14 above)
Comment (Johnetordoff) [reply api_tests/revisions/views/test_registrations_schema_responses_list.py:41]: I am going off the API contract, so not sure what is correct.

File: api/registrations/views.py [hunk 20]
```diff
@@ -850,3 +858,26 @@ def get_default_queryset(self):
 
     def get_queryset(self):
         return self.get_queryset_from_request()
+
+
+class RegistrationSchemaResponsesList(JSONAPIBaseView, ListFilterMixin, generics.ListCreateAPIView, RegistrationMixin):
```
Comment (jwalz) [inline api/registrations/views.py:863]: no need to post here

File: api_tests/revisions/views/test_schema_response_list.py (hunk 12 above)
Comment (Johnetordoff) [reply api_tests/revisions/views/test_schema_response_list.py]: Nope removed

File: api/revisions/serializers.py (hunk 11 above)
Comment (Johnetordoff) [reply api/revisions/serializers.py]: removed

File: api/revisions/serializers.py [hunk 21]
```diff
@@ -0,0 +1,155 @@
...
+    reviews_state = ser.ChoiceField(choices=['revision_in_progress', 'revision_pending_admin_approval', 'revision_pending_moderation', 'approved'], required=False)
+    is_pending_current_user_approval = ser.SerializerMethodField()
+
+    links = LinksField(
+        {
+            'self': 'get_absolute_url',
+        },
+    )
+
+    registration = RelationshipField(
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply api/revisions/serializers.py]: This is ultimately what I went this.
```
    parent = GenericRelationshipField(
        related_view=(
            lambda object: {
                'osf.registration': 'registrations:registration-detail',
                'osf.node': 'nodes:node-detail',
                'osf.preprint': 'preprints:preprint-detail',
            }[object.parent.type]
        ),
        related_view_kwargs=(
            lambda object: {
                'osf.registration': {'node_id': '<parent._id>'},
                'osf.node': {'node_id': '<parent._id>'},
                'osf.preprint': {'preprint_id': '<parent._id>'},
            }[object.parent.type]
        ),
        required=False,
    )
```

File: api/base/serializers.py [hunk 22]
```diff
@@ -1213,6 +1213,36 @@ def _handle_callable_view(self, obj, view):
         return view(getattr(obj, self.view_lambda_argument))
 
 
+class GenericRelationshipField(RelationshipField):
```
Comment (jwalz) [inline api/base/serializers.py:1216]: Should we just be re-using the `TargetField` that Files use to resolve their GenericForeignKey? https://github.com/CenterForOpenScience/osf.io/blob/2ca525a761d2b27cc7f244226ea55166fa4ea9b2/api/base/serializers.py#L936

File: osf_tests/factories.py [hunk 23]
```diff
@@ -1115,3 +1115,81 @@ class Meta:
...
+        initiator = kwargs.get('initiator')
+        parent = kwargs.get('parent')
+        schema = kwargs.get('schema')
+        parent.registered_schema.add(schema)
+
+        if SchemaResponse.objects.filter(object_id=parent.id, content_type_id=ContentType.objects.get_for_model(parent)).exists():
+            previous_schema_response = SchemaResponse.objects.filter(
+                object_id=parent.id,
+                content_type_id=ContentType.objects.get_for_model(parent)
+            ).order_by('created').first()
```
Comment (jwalz) [inline osf_tests/factories.py:1192]: `order_by` defaults to ascending order. This should either be `order_by('-created').first()` or `order_by('created').last()`. Could also use `updated_response__isnull=True` in the filter to get the latest un-revised response (assuming we're not putting ourselves into. weird state).

**Review by jwalz — CHANGES_REQUESTED** [review]
A couple of lingering questions around API stuff and a few more around tests.

File: api/base/serializers.py (hunk 22 above)
Comment (Johnetordoff) [reply api/base/serializers.py:1216]: The TargetField relies on the resource having a guid referent, I'm actually thinking of removing all these TargetField stuff, because it's a big class and it's only used 2 places and the `GenericRelationshipField` is move explicit and flexible. Any thoughts? Honestly it looks like a lot of these fields could be replaced with GenericRelationshipFields.

File: api/schema_responses/serializers.py [hunk 24]
```diff
@@ -0,0 +1,124 @@
...
+
+    def get_is_pending_current_user_approval(self, obj):
+        # TBD
+        return False
+
+    def create(self, validated_data):
+        registration = Registration.load(validated_data.pop('_id'))
+
+        try:
+            schema = registration.registered_schema.get()
```
Comment (jwalz) [inline api/schema_responses/serializers.py:93]: looks like the Registration already has a property to grab this: [registration_schema](https://github.com/CenterForOpenScience/osf.io/blob/develop/osf/models/registrations.py#L198)

File: api/schema_responses/serializers.py [hunk 25]
```diff
@@ -0,0 +1,124 @@
...
+    registration = RelationshipField(
+        related_view='registrations:registration-detail',
+        related_view_kwargs={'node_id': '<parent._id>'},
+        read_only=True,
+        required=False,
+    )
+
+    registration_schema = RelationshipField(
+        related_view='schemas:registration-schema-detail',
+        related_view_kwargs={'schema_id': '<parent.registered_schema_id>'},
```
Comment (jwalz) [inline api/schema_responses/serializers.py:60]: Just use self.schema, please, that will help us avoid weirdness with the the registration using a many-to-many field

File: api_tests/schema_responses/views/test_schema_response_detail.py [hunk 26]
```diff
@@ -0,0 +1,154 @@
...
+
+        revised_schema.update_responses({'q1': {'value': 'update value'}, 'q2': {}})
+        resp = app.get(f'/v2/schema_responses/{revised_schema._id}/', auth=user.auth)
+        assert resp.status_code == 200
+        data = resp.json['data']
+        assert data['id'] == revised_schema._id
+        assert data['attributes']['updated_response_keys'] == ['q1']
+
+    def test_schema_response_detail_validation(self, app, schema_response, invalid_payload, user, url):
+        schema_response.parent.add_contributor(user, 'admin')
```
Comment (jwalz) [inline api_tests/schema_responses/views/test_schema_response_detail.py:133]: why are we adding this user here?

File: api_tests/schema_responses/views/test_registrations_schema_responses_list.py [hunk 27]
```diff
@@ -0,0 +1,53 @@
...
+            initiator=registration.creator,
+            revision_justification="We're talkin' about practice.",
+        )
+
+    @pytest.fixture()
+    def url(self, registration):
+        return f'/v2/registrations/{registration._id}/schema_responses/'
+
+    def test_registrations_schema_responses_list(self, app, schema_response, schema_response2, user, url):
+        resp = app.get(url, auth=user.auth, expect_errors=True)
```
Comment (jwalz) [inline api_tests/schema_responses/views/test_registrations_schema_responses_list.py:42]: can you make this 403 case its own test?

File: api/schema_responses/serializers.py [hunk 28]
```diff
@@ -0,0 +1,124 @@
...
+            schema_response = SchemaResponse.create_from_previous_response(
+                initiator=initiator,
+                previous_response=registration.schema_responses.order_by('-created').first(),
+                justification=justification,
+            )
+
+        return schema_response
+
+    def update(self, schema_response, validated_data):
+        schema_responses = validated_data.get('revision_response')
```
Comment (jwalz) [inline api/schema_responses/serializers.py:117]: should be `revision_responses`, right?

File: api_tests/schema_responses/views/test_schema_response_detail.py [hunk 29]
```diff
@@ -0,0 +1,154 @@
...
+    RegistrationFactory,
+    AuthUserFactory
+)
+
+from osf.models import SchemaResponse
+from osf_tests.utils import DEFAULT_TEST_SCHEMA
+
+
+@pytest.mark.django_db
+class TestSchemaResponseDetail:
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply api_tests/schema_responses/views/test_schema_response_detail.py:14]: How do you want to do these auth tests? Since it's only auth we can write really parameterized test. I would like to have these be very declarative. We should brainstorm this.

File: api/schema_responses/views.py [hunk 30]
```diff
@@ -0,0 +1,50 @@
...
+
+from api.schema_responses.serializers import (
+    SchemaResponseRegistrationParentSerializer,
+)
+from osf.models import SchemaResponse, Registration
+from api.base.filters import ListFilterMixin
+
+
+class SchemaResponseList(JSONAPIBaseView, ListFilterMixin, generics.ListAPIView):
+    permission_classes = (
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply api/schema_responses/views.py:14]: Protecting against unsupported methods is done automatically by the view, so the fact that it's a `ListAPIView` and not a `ListCreateAPIView` is going to mean POSTs get a 405.

[the remark this answers is not in the document]
Comment (jwalz) [reply api/schema_responses/views.py:14]: This needs the create view. Right now, no endpoint supports creating a SchemaResponse.
